### PR TITLE
ROOT 6.32.00 prerequisites

### PR DESCRIPTION
**Pull request by Barthelemy**

### CODE REVIEW

**Comment by Barthelemy** [comment]
As discussed with @ktf , it solves the issue we see with RPMs and it contains a fix needed by @knopers8

**Comment by ktf** [comment]
@ihrivnac any idea of why we get:

```
/sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx: In member function 'virtual VGM::ThreeVector RootGM::TessellatedSolid::Vertex(int, int) const':
/sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx:184:26: error: 'const class TGeoFacet' has no member named 'GetVertex'
  184 |   vertex.push_back(facet.GetVertex(index).fVec[0] * RootGM::Units::Length());
```

?

**Comment by Barthelemy** [comment]
In O2 (with dataflow defaults) we see
```
2024-06-12@15:01:03:DEBUG:O2Suite:O2:0: /home/alibuild/sw/SOURCES/O2/daily-20240607-0200/daily-20240607-0200/Framework/Core/include/Framework/HistogramRegistry.h:300:76: error: 'TArrayL' was not declared in this scope; did you mean 'TArrayS'?
2024-06-12@15:01:03:DEBUG:O2Suite:O2:0:   300 |     return getBaseElementSize<TArrayD, TArrayF, TArrayC, TArrayI, TArrayC, TArrayL>(ptr);
2024-06-12@15:01:03:DEBUG:O2Suite:O2:0:       |                                                                            ^~~~~~~
2024-06-12@15:01:03:DEBUG:O2Suite:O2:0:       |                                                                            TArrayS
```
I presume that `TArrayL` has been removed. I will ask ROOT.

**Comment by Barthelemy** [comment]
well no, `TArrayL` is there.

File: root.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package: ROOT
 version: "%(tag_basename)s"
-tag: "v6-30-01-alice5"
-source: https://github.com/alisw/root.git
+tag: "v6-32-00"
+source: https://github.com/root-project/root.git
```
Comment (sawenzel) [inline root.sh:4]: Note that we have custom commits in our ALICE specific branches that would be skipped with this move. 

Also, we need to make sure that any bump in ROOT contains the fix for this issue: https://github.com/root-project/root/issues/15579.

Comment (Barthelemy) [reply root.sh:4]: ok, then I guess that we need to get 6.32.00 inside `alisw` and apply the patches. @ktf ?

Comment (ktf) [inline root.sh:4]:
```suggestion
tag: "v6-32-00-alice1"
source: https://github.com/alisw/root.git
```

Comment (ktf) [reply root.sh]: I have forked alice/v6-32-00 and added the remaining patches from v6-30-01-alice5. @sawenzel can you open a PR to the alisw/root branch with the extra commits you need?

**Comment by ktf** [comment]
For the issue with JAliEn-ROOT we need https://gitlab.cern.ch/jalien/jalien-root/-/merge_requests/85

**Comment by ktf** [comment]
@costing can you check and eventually tag?

**Comment by costing** [comment]
Thank you @ktf , JAliEn-ROOT tagged as 0.7.12

**Comment by ktf** [comment]
@costing thanks. @Barthelemy I have added the new JAliEn-ROOT to your branch.

**Comment by sawenzel** [comment]
> @ihrivnac any idea of why we get:
> 
> ```
> /sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx: In member function 'virtual VGM::ThreeVector RootGM::TessellatedSolid::Vertex(int, int) const':
> /sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx:184:26: error: 'const class TGeoFacet' has no member named 'GetVertex'
>   184 |   vertex.push_back(facet.GetVertex(index).fVec[0] * RootGM::Units::Length());
> ```
> 
> ?

ROOT unfortunately changed some TGeo API. To move forward, we either need to take back that change or we need to adapt our software stack (VGM). @ihrivnac : Are you already working on this?

**Comment by sawenzel** [comment]
I see that this has already been fixed in VGM via https://github.com/vmc-project/vgm/commit/bb570ae367aeca90908862d6912265e769bfa50e. So all we need is a new tag.

**Comment by ktf** [comment]
@ihrivnac can you take care of the new tag?

**Comment by ktf** [comment]
HepMC seems to have troubles as well.

```
[01m[K/sw/slc8_x86-64/ROOT/v6-32-00-alice1-local1/include/TFile.h:332:45:[m[K [01;31m[Kerror: [m[K'[01m[Kstd::string_view[m[K' has not been declared
  332 |    static Bool_t       SetCacheFileDir(std::[01;31m[Kstring_view[m[K cacheDir, Bool_t operateDisconnected = kTRUE,
      |                                             [01;31m[K^~~~~~~~~~~[m[K
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:162: rootIO/CMakeFiles/HepMC3rootIO.dir/G__HepMC3rootIO.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[01m[K/sw/slc8_x86-64/ROOT/v6-32-00-alice1-local1/include/TFile.h:332:45:[m[K [01;31m[Kerror: [m[K'[01m[Kstd::string_view[m[K' has not been declared
  332 |    static Bool_t       SetCacheFileDir(std::[01;31m[Kstring_view[m[K cacheDir, Bool_t operateDisconnected = kTRUE,
      |                                             [01;31m[K^~~~~~~~~~~[m[K
--
[01m[K/sw/slc8_x86-64/ROOT/v6-32-00-alice1-local1/include/TDirectory.h:269:24:[m[K [01;36m[Knote: [m[Kdeclared private here
  269 |            Int_t       [01;36m[KWriteObject[m[K(void *obj, const char* name, Option_t *option="", Int_t bufsize=0); // Intentionally not implemented.
      |                        [01;36m[K^~~~~~~~~~~[m[K
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:148: rootIO/CMakeFiles/HepMC3rootIO.dir/src/Streamers.cc.o] Error 1
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:106: rootIO/CMakeFiles/HepMC3rootIO.dir/src/ReaderRoot.cc.o] Error 1
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:92: rootIO/CMakeFiles/HepMC3rootIO.dir/src/WriterRoot.cc.o] Error 1
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:120: rootIO/CMakeFiles/HepMC3rootIO.dir/src/WriterRootTree.cc.o] Error 1
make[2]: *** [rootIO/CMakeFiles/HepMC3rootIO.dir/build.make:134: rootIO/CMakeFiles/HepMC3rootIO.dir/src/ReaderRootTree.cc.o] Error 1
make[2]: Leaving directory '/sw/BUILD/c5792512cdafbc608dd346d36769f9c16565a610/HepMC3'
```

**Comment by ihrivnac** [comment]
> > @ihrivnac any idea of why we get:
> > ```
> > /sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx: In member function 'virtual VGM::ThreeVector RootGM::TessellatedSolid::Vertex(int, int) const':
> > /sw/SOURCES/vgm/v5-2/v5-2/packages/RootGM/source/solids/TessellatedSolid.cxx:184:26: error: 'const class TGeoFacet' has no member named 'GetVertex'
> >   184 |   vertex.push_back(facet.GetVertex(index).fVec[0] * RootGM::Units::Length());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ?
> 
> ROOT unfortunately changed some TGeo API. To move forward, we either need to take back that change or we need to adapt our software stack (VGM). @ihrivnac : Are you already working on this?

The new vgm tag (v5-3) is now in place

**Comment by ktf** [comment]
@sawenzel I have bumped VGM to v5-3 and HepMC to 3.3.0 which seems to work. I now get an issue with pythia6 support being dropped in ROOT and geant3 failing because of that. Can you remind me what we decided for this?

**Comment by ihrivnac** [comment]
> @sawenzel I have bumped VGM to v5-3 and HepMC to 3.3.0 which seems to work. I now get an issue with pythia6 support being dropped in ROOT and geant3 failing because of that. Can you remind me what we decided for this?

In the last geant3 tag, we removed the dependency on pythia6. Can you give more details about this geant3 failing ?

**Comment by ktf** [comment]
@ihrivnac The issue I saw in GEANT3 was actually due to my old tag in the development package. Sorry about the noise.

However there is an actual issue in AEGIS, apparently:

https://ali-ci.cern.ch/alice-build-logs/alisw/alidist/5502/43a6cdfc153366595adcc670603df6ce34424eaa/build_O2_fullCI_alidist/pretty.html

**Comment by ktf** [comment]
Rolling this back to v6-30-01-alice5 so that we can at least merge the prerequisites from the new ROOT, without actually bumping ROOT itself.

**Comment by Barthelemy** [comment]
@ktf will you merge this PR to have the prerequisite and open another one for 6.32 ?

**Comment by ktf** [comment]
The problem is that the errors are genuine (see [1] for the Rivet one, [2] for AliRoot, [3] for AEGIS) due to the changes in HepMC / VGM which we do need for the new ROOT. @ihrivnac @sawenzel @pzhristov @cholmcc could you have a look?


- [1]:
```
## sw/BUILD/Rivet-latest/log
--
                 from /sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/ReaderFactory_fwd.h:19,
                 from /sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/ReaderFactory.h:15,
                 from RivetHepMC_3.cc:7:
/sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/bxzstr/strict_fstream.hpp:57:7: error: redefinition of 'class strict_fstream::Exception'
   57 | class Exception
      |       ^~~~~~~~~
In file included from ../Core/zstr/zstr.hpp:16,
--
../Core/zstr/strict_fstream.hpp:49:7: note: previous definition of 'class strict_fstream::Exception'
   49 | class Exception
      |       ^~~~~~~~~
/sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/bxzstr/strict_fstream.hpp:70:8: error: redefinition of 'struct strict_fstream::detail::static_method_holder'
   70 | struct static_method_holder
      |        ^~~~~~~~~~~~~~~~~~~~
../Core/zstr/strict_fstream.hpp:62:8: note: previous definition of 'struct strict_fstream::detail::static_method_holder'
--
../Core/zstr/strict_fstream.hpp:62:8: note: previous definition of 'struct strict_fstream::detail::static_method_holder'
   62 | struct static_method_holder
      |        ^~~~~~~~~~~~~~~~~~~~
/sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/bxzstr/strict_fstream.hpp:148:7: error: redefinition of 'class strict_fstream::ifstream'
  148 | class ifstream
      |       ^~~~~~~~
../Core/zstr/strict_fstream.hpp:143:7: note: previous definition of 'class strict_fstream::ifstream'
--
../Core/zstr/strict_fstream.hpp:143:7: note: previous definition of 'class strict_fstream::ifstream'
  143 | class ifstream
      |       ^~~~~~~~
/sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/bxzstr/strict_fstream.hpp:168:7: error: redefinition of 'class strict_fstream::ofstream'
  168 | class ofstream
      |       ^~~~~~~~
../Core/zstr/strict_fstream.hpp:163:7: note: previous definition of 'class strict_fstream::ofstream'
--
../Core/zstr/strict_fstream.hpp:163:7: note: previous definition of 'class strict_fstream::ofstream'
  163 | class ofstream
      |       ^~~~~~~~
/sw/slc7_x86-64/HepMC3/3.3.0-local3/include/HepMC3/bxzstr/strict_fstream.hpp:187:7: error: redefinition of 'class strict_fstream::fstream'
  187 | class fstream
      |       ^~~~~~~
../Core/zstr/strict_fstream.hpp:182:7: note: previous definition of 'class strict_fstream::fstream'
--
  182 | class fstream
      |       ^~~~~~~
mv -f .deps/libRivetTools_la-ParticleName.Tpo .deps/libRivetTools_la-ParticleName.Plo
make[2]: *** [libRivetTools_la-RivetHepMC_3.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/libRivetTools_la-RivetMT2.Tpo .deps/libRivetTools_la-RivetMT2.Plo
mv -f .deps/libRivetTools_la-Vector4.Tpo .deps/libRivetTools_la-Vector4.Plo
mv -f .deps/libRivetTools_la-Random.Tpo .deps/libRivetTools_la-Random.Plo
--
mv -f .deps/libRivetTools_la-Correlators.Tpo .deps/libRivetTools_la-Correlators.Plo
mv -f .deps/libRivetTools_la-RivetYODA.Tpo .deps/libRivetTools_la-RivetYODA.Plo
make[2]: Leaving directory `/sw/BUILD/0f0ee3d44bc4ab3ddc1543ee42417ddeea96c860/Rivet/src/Tools'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/sw/BUILD/0f0ee3d44bc4ab3ddc1543ee42417ddeea96c860/Rivet/src'
make: *** [all-recursive] Error 1
```

- [2]:
```
[ 77%] Building CXX object TRD/TRDbase/CMakeFiles/TRDbase-object.dir/AliTRDSimParam.cxx.o
[ 77%] Building CXX object EVGEN/CMakeFiles/EVGEN.dir/AliGenReadersEMD.cxx.o
[ 77%] Generating libMUONtrigger.rootmap, G__MUONtrigger.cxx, G__MUONtrigger_rdict.pcm
/sw/SOURCES/AliRoot/build_AliRoot_alidist-slc7_x86-64/0/EVGEN/AliGenPerformance.cxx:66:10: fatal error: TMCParticle.h: No such file or directory
   66 | #include "TMCParticle.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.
--
   66 | #include "TMCParticle.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [EVGEN/CMakeFiles/EVGEN.dir/AliGenPerformance.cxx.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
[ 77%] Building CXX object TRD/TRDbase/CMakeFiles/TRDbase-object.dir/AliTRDtrackGTU.cxx.o
[ 77%] Building CXX object PMD/PMDrec/CMakeFiles/PMDrec-object.dir/AliPMDReconstructor.cxx.o
[ 77%] Linking CXX shared library ../../lib/libVZEROrec.so
```

- [3]:
```
## sw/BUILD/AEGIS-latest/log
--
[ 43%] Generating G__TEPEMGEN.cxx, libTEPEMGEN_rdict.pcm, libTEPEMGEN.rootmap
In file included from input_line_9:3:
In file included from /sw/SOURCES/AEGIS/v1.5.3/v1.5.3/GeneratorParam/./GeneratorParam.h:19:
/sw/SOURCES/AEGIS/v1.5.3/v1.5.3/GeneratorParam/./PythiaDecayerConfig.h:22:10: fatal error: 'TPythia6.h' file not found
#include <TPythia6.h>
         ^~~~~~~~~~~~
In file included from input_line_9:3:
--
/sw/slc8_x86-64/ROOT/v6-30-01-alice5-local15/include/Rtypes.h:314:25: note: expanded from macro '_ClassDefOutline_'
   virtual_keyword void Streamer(TBuffer&) overrd;
                        ^
```

**Comment by sawenzel** [comment]
Fixing the TPythia6 error in AEGIS appears to be non-trivial. I have opened a ticket here to follow up: https://its.cern.ch/jira/browse/O2-5100. We could, for now, also copy the old TPythia6 code to AEGIS for a quick solution.

**Comment by ktf** [comment]
@sawenzel ok. I copied the header in v1.5.3-alice1 and that allows me to continue. Is AEGIS actually our own code?

**Comment by sawenzel** [comment]
> @sawenzel ok. I copied the header in v1.5.3-alice1 and that allows me to continue. Is AEGIS actually our own code?

Yes AEGIS is our own code. I am not quite sure which header (wouldn't we also need source files?) you copied but it would be fine with me going this direction.

**Comment by ktf** [comment]
I copied all the "TPythia6*.h" and that is enough to compile it. Most likely they do not need anything from the cxx for that (it compiles without resolving symbols).

**Comment by Barthelemy** [comment]
Hi, 
Where do we stand ?
What is missing ? 
Thank you

**Comment by ktf** [comment]
@sawenzel fixed AEGIS for good yesterday. The remaining bit is Rivet, AFACIT. If that only affects AliGenerators, I would go ahead and merge this.

**Comment by ktf** [comment]
https://github.com/alisw/ali-bot/pull/1363 will make sure we use the old ROOT (i.e. defaults-ali) for the alidist test of AliRoot as well (and not only the release).

**Pull request merged**